### ISSUE TITLE
Don't allow requires_grad=True on Tensors with integral types

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#3226]
1. We should add a check when setting `requires_grad=True` that the tensor is floating point
2. We need to fix the gradcheck tests so that they don't try setting `requires_grad=True` on integer types

**Comment by colesbury** [comment]
I think @gchanan fixed this

**Comment by AdityaKansal** [comment]
It would be great help if someone can please help or guide me through a link to understand why we are not allowing integers value while calculating gradients? As per my understanding, once we have the d/dx of  some function, then we should be able to calculate its value at integer value of x as well.

**Comment by colesbury** [comment]
> It would be great help if someone can please help or guide me through a link to understand why we are not allowing integers value while calculating gradients

We allow integer values. Floating-point tensors can represent integer values. You can even compute derivatives of functions that are integer-valued (floor(), ceil(), round()), although the derivative will be zero, which might not be useful for optimization.

We don't allow derivatives of functions where the **domain of the function** is limited to integers. That's same as functions on tensors with integral types. Derivatives aren't [defined](https://simple.wikipedia.org/wiki/Derivative_(mathematics)) on these functions and even extensions like [subderivatives](https://en.wikipedia.org/wiki/Subderivative) don't make sense here: you can't compute `lim ε -> 0 [(f(x + ε) - f(x)) / ε]` if `f` is only defined on integers.

**Comment by AdityaKansal** [comment]
Thank you so much your detailed answer.
